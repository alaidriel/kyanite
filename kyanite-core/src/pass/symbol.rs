use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    ops::{Deref, DerefMut},
};

use crate::{
    ast::{node, Decl, Type},
    token::Token,
};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Binding {
    Function(node::FuncDecl),
    Constant(node::ConstantDecl),
    Variable(node::VarDecl),
}

impl Binding {
    pub fn ty(&self) -> Type {
        match self {
            Self::Function(fun) => Type::from(fun.ty.as_ref()),
            Self::Constant(c) => Type::from(&c.ty),
            Self::Variable(v) => Type::from(&v.ty),
        }
    }
}

#[derive(Debug)]
pub struct SymbolTable(HashMap<Token, Binding>);

impl SymbolTable {
    pub fn new() -> Self {
        Self(HashMap::new())
    }
}

impl Deref for SymbolTable {
    type Target = HashMap<Token, Binding>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for SymbolTable {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl Default for SymbolTable {
    fn default() -> Self {
        Self::new()
    }
}

impl From<&Vec<Decl>> for SymbolTable {
    fn from(nodes: &Vec<Decl>) -> Self {
        let mut table = Self::new();
        for node in nodes {
            self::SymbolTableVisitor::visit(node, &mut table);
        }
        table
    }
}

trait SymbolTableVisitor {
    fn visit(&self, table: &mut SymbolTable);
}

impl SymbolTableVisitor for Decl {
    fn visit(&self, table: &mut SymbolTable) {
        match self {
            Decl::Function(fun) => func(fun, table),
            Decl::Constant(c) => constant(c, table),
        }
    }
}

fn func(fun: &node::FuncDecl, table: &mut SymbolTable) {
    table.insert(fun.name.clone(), Binding::Function(fun.clone()));
}

fn constant(c: &node::ConstantDecl, table: &mut SymbolTable) {
    table.insert(c.name.clone(), Binding::Constant(c.clone()));
}
