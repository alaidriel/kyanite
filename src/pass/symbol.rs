use std::collections::HashMap;

use crate::ast::{node, File, Node, Type};

#[derive(Debug)]
pub struct Function {
    pub(super) arity: usize,
    pub(super) ty: Type,
    pub(super) params: Vec<(String, Type)>,
}

impl Function {
    pub fn new(arity: usize, ty: Type, params: Vec<(String, Type)>) -> Self {
        Self { arity, ty, params }
    }
}

#[derive(Debug)]
pub enum Symbol {
    Function(Function),
    Variable(Type),
}

impl Symbol {
    pub fn ty(&self) -> Type {
        match self {
            Symbol::Function(f) => f.ty.clone(),
            Symbol::Variable(ty) => ty.clone(),
        }
    }
}

#[derive(Debug)]
pub struct SymbolTable(HashMap<String, Symbol>);

impl SymbolTable {
    pub fn new() -> Self {
        Self(HashMap::new())
    }

    pub fn get(&self, name: &String) -> Option<&Symbol> {
        self.0.get(name)
    }

    pub fn insert(&mut self, name: String, symbol: Symbol) {
        self.0.insert(name, symbol);
    }
}

impl From<Vec<(String, Symbol)>> for SymbolTable {
    fn from(vars: Vec<(String, Symbol)>) -> Self {
        Self(vars.into_iter().collect())
    }
}

impl From<&File> for SymbolTable {
    fn from(program: &File) -> Self {
        let mut table = Self::new();
        for node in &program.nodes {
            self::SymbolTableVisitor::visit(node, &mut table);
        }
        // dbg!(&table);
        table
    }
}

trait SymbolTableVisitor {
    fn visit(&self, table: &mut SymbolTable);
}

impl SymbolTableVisitor for Node {
    fn visit(&self, table: &mut SymbolTable) {
        match self {
            Node::FuncDecl(fun) => func(fun, table),
            Node::ConstantDecl(c) => constant(c, table),
            _ => {}
        }
    }
}

fn func(fun: &node::FuncDecl, table: &mut SymbolTable) {
    let params: Vec<(String, Type)> = fun
        .params
        .iter()
        .map(|param| (String::from(&param.name), Type::from(&param.ty)))
        .collect();
    table.insert(
        String::from(&fun.name),
        Symbol::Function(Function::new(
            fun.params.len(),
            Type::from(fun.ty.as_ref()),
            params,
        )),
    );
}

fn constant(c: &node::ConstantDecl, table: &mut SymbolTable) {
    table.insert(String::from(&c.name), Symbol::Variable(Type::from(&c.ty)));
}
