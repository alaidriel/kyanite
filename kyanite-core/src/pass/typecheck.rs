use std::ops::Deref;

use crate::{
    ast::{node, File, Node, Type},
    token::Token,
};

use super::symbol::{Symbol, SymbolTable};

pub struct TypeCheckPass<'a> {
    program: &'a File,
    errors: usize,
    function: Option<String>,
    scopes: Vec<SymbolTable>,
}

impl<'a> TypeCheckPass<'a> {
    pub fn new(table: SymbolTable, program: &'a File) -> Self {
        Self {
            program,
            errors: 0,
            function: None,
            scopes: vec![table],
        }
    }

    pub fn run(&mut self) -> usize {
        for node in &self.program.nodes {
            self.check(node);
        }
        self.errors
    }

    fn scope_mut(&mut self) -> &mut SymbolTable {
        self.scopes.last_mut().unwrap()
    }

    fn begin_scope(&mut self) {
        self.scopes.push(SymbolTable::new());
    }

    fn end_scope(&mut self) {
        self.scopes.pop();
    }

    fn symbol(&self, name: &String) -> Option<&Symbol> {
        for scope in self.scopes.iter().rev() {
            if let Some(symbol) = scope.get(name) {
                return Some(symbol);
            }
        }
        None
    }

    fn error(&mut self, at: &Token, msg: String) {
        eprintln!("{} at {}", msg, at.span);
        self.errors += 1;
    }

    fn check(&mut self, node: &Node) -> Type {
        match node {
            Node::Float(_) => Type::Float,
            Node::Int(_) => Type::Int,
            Node::Bool(_) => Type::Bool,
            Node::Str(_) => Type::Str,
            Node::ConstantDecl(c) => self.constant(c),
            Node::VarDecl(v) => self.var(v),
            Node::FuncDecl(fun) => self.function(fun),
            Node::Return(r) => self.ret(r),
            Node::Binary(b) => self.binary(b),
            Node::Ident(id) => self.ident(id),
            Node::Call(call) => self.call(call),
            e => todo!("{:?}", e),
        }
    }

    fn constant(&mut self, c: &node::ConstantDecl) -> Type {
        let got = self.check(&c.expr);
        let expected = Type::from(&c.ty);
        if got != expected {
            self.error(
                &c.ty,
                format!(
                    "Mismatched types in constant declaration (expected {:?}, got {:?})",
                    Type::from(&c.ty),
                    got,
                ),
            );
        }
        expected
    }

    fn var(&mut self, v: &node::VarDecl) -> Type {
        let got = self.check(&v.expr);
        let expected = Type::from(&v.ty);
        if got != expected {
            self.error(
                &v.ty,
                format!(
                    "Mismatched types in variable declaration (expected {:?}, got {:?})",
                    Type::from(&v.ty),
                    got,
                ),
            );
        }
        self.scope_mut()
            .insert(String::from(&v.name), Symbol::Variable(Type::from(&v.ty)));
        expected
    }

    fn function(&mut self, fun: &node::FuncDecl) -> Type {
        self.begin_scope();
        self.function = Some(String::from(&fun.name));
        for param in &fun.params {
            self.scope_mut().insert(
                String::from(&param.name),
                Symbol::Variable(Type::from(&param.ty)),
            );
        }
        for node in &fun.body {
            self.check(node);
        }
        self.end_scope();
        self.function = None;
        Type::Void
    }

    fn ret(&mut self, r: &node::Return) -> Type {
        let got = self.check(&r.expr);
        match &self.function {
            Some(function) => {
                let symbol = self.symbol(function).unwrap();
                if got != symbol.ty() {
                    self.error(
                        &r.keyword,
                        format!(
                            "Mismatched types in return (expected {:?}, got {:?})",
                            symbol.ty(),
                            got
                        ),
                    );
                }
            }
            None => self.error(&r.keyword, "Return outside of function".into()),
        }
        got
    }

    fn binary(&mut self, b: &node::Binary) -> Type {
        let lty = self.check(&b.left);
        let rty = self.check(&b.right);
        if lty != rty {
            self.error(
                &b.op,
                format!(
                    "Mismatched types in '{}' operation ({:?}, {:?})",
                    b.op, lty, rty
                ),
            );
        }
        lty
    }

    fn ident(&mut self, id: &node::Ident) -> Type {
        let name = String::from(&id.name);
        match self.symbol(&name) {
            Some(Symbol::Variable(ty)) => ty.clone(),
            _ => {
                self.error(&id.name, format!("{} is not defined", name));
                Type::Void
            }
        }
    }

    fn call(&mut self, call: &node::Call) -> Type {
        let name = match &call.left.deref() {
            Node::Ident(ident) => String::from(&ident.name),
            Node::Binary(_) => todo!(),
            _ => unimplemented!(),
        };
        let (arity, params, ty) = match self.symbol(&name) {
            Some(Symbol::Function(function)) => {
                (function.arity, function.params.clone(), function.ty.clone())
            }
            Some(_) => {
                self.error(&call.parens.0, format!("{} is not a function", name));
                return Type::Void;
            }
            _ => {
                // TODO: more robust type checking for builtins
                if name != "println" && name != "max" && name != "min" {
                    self.error(&call.parens.0, format!("{} is not defined", name));
                }
                return Type::Void;
            }
        };
        if arity != call.args.len() {
            self.error(
                &call.parens.1,
                format!("Expected {} arguments, got {}", arity, call.args.len()),
            );
        }
        for (i, arg) in call.args.iter().enumerate() {
            let ty = self.check(arg);
            if i < params.len() && ty != params[i].1 {
                let token = match i {
                    0 => &call.parens.0,
                    _ if i == params.len() - 1 => &call.parens.1,
                    _ => &call.delimiters[i],
                };
                self.error(
                    token,
                    format!(
                        "Mismatched type in function parameter: (expected {:?}, got {:?})",
                        params[i].1, ty
                    ),
                );
            }
        }
        ty.clone()
    }
}
