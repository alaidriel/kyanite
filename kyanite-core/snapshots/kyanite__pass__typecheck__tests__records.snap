---
source: kyanite-core/src/pass/typecheck.rs
expression: pass.errors
---
- filename: test-cases/typecheck/records.kya
  heading: "no field `hello` on type `Person`"
  source: "        hello: \"world\""
  span:
    line: 22
    column: 9
    length: 5
  text: ""
- filename: test-cases/typecheck/records.kya
  heading: expected expression of type Person
  source: "    alaina = 5;"
  span:
    line: 25
    column: 14
    length: 1
  text: expression of type int
- filename: test-cases/typecheck/records.kya
  heading: expected expression of type bool
  source: "    alaina.foo.bar.baz = Bar:init(baz: false);"
  span:
    line: 28
    column: 26
    length: 20
  text: expression of type Bar
- filename: test-cases/typecheck/records.kya
  heading: expected initializer to be of type int
  source: "    let name: int = alaina.name;"
  span:
    line: 30
    column: 21
    length: 11
  text: expression of type str
- filename: test-cases/typecheck/records.kya
  heading: "no field `faavorite` on type `Person`"
  source: "    let favorite: int = alaina.faavorite;"
  span:
    line: 31
    column: 32
    length: 9
  text: ""
- filename: test-cases/typecheck/records.kya
  heading: "no field `barr` on type `Foo`"
  source: "    let bar: Bar = alaina.foo.barr;"
  span:
    line: 33
    column: 31
    length: 4
  text: ""
- filename: test-cases/typecheck/records.kya
  heading: "no field `baaz` on type `Bar`"
  source: "    let baz: Baz = alaina.foo.bar.baaz;"
  span:
    line: 34
    column: 35
    length: 4
  text: ""
- filename: test-cases/typecheck/records.kya
  heading: expected initializer to be of type Baz
  source: "    let baz: Baz = alaina.foo.bar;"
  span:
    line: 37
    column: 20
    length: 14
  text: expression of type Bar
- filename: test-cases/typecheck/records.kya
  heading: expected initializer to be of type Baz
  source: "    let baz: Baz = alaina.foo.bar.baz;"
  span:
    line: 38
    column: 20
    length: 18
  text: expression of type bool
- filename: test-cases/typecheck/records.kya
  heading: "`foo` is not defined"
  source: "    foo.baz;"
  span:
    line: 40
    column: 5
    length: 3
  text: ""

